src/oidc_child/oidc_child_json.c

```diff
@@ -0,0 +1,505 @@
+    if (dc_ctx->user_code != NULL) {
+        talloc_set_destructor((void *) dc_ctx->user_code, sss_erase_talloc_mem_securely);
+    }
+    dc_ctx->device_code = get_json_string(dc_ctx, root, "device_code");
+    if (dc_ctx->device_code != NULL) {
+        talloc_set_destructor((void *) dc_ctx->device_code, sss_erase_talloc_mem_securely);
+    }
+    dc_ctx->verification_uri = get_json_string(dc_ctx, root,
+                                               "verification_uri");
+    if (dc_ctx->verification_uri == NULL) {
+        /* Google uses _urL rather than _urI, see e.g.
+         * https://developers.google.com/identity/protocols/oauth2/limited-input-device
+         * Old Azure AD v1 endpoints do the same. */
+        dc_ctx->verification_uri = get_json_string(dc_ctx, root,
+                                                   "verification_url");
+    }
+    dc_ctx->verification_uri_complete = get_json_string(dc_ctx, root,
+                                                   "verification_uri_complete");
+    dc_ctx->message = get_json_string(dc_ctx, root, "message");
+    dc_ctx->interval = get_json_integer(root, "interval");
+    dc_ctx->expires_in = get_json_integer(root, "expires_in");
+
+    ret = EOK;
+
+done:
+    json_decref(root);
+    return ret;
+}
+
+errno_t parse_token_result(struct devicecode_ctx *dc_ctx,
+                           char **error_description)
+{
+    json_t *tmp = NULL;
+    json_error_t json_error;
+
+    *error_description = NULL;
+    dc_ctx->result = json_loads(dc_ctx->http_data, 0, &json_error);
+    if (dc_ctx->result == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to parse json data on line [%d]: [%s].\n",
+              json_error.line, json_error.text);
+        return EINVAL;
+    }
+
+    tmp = json_object_get(dc_ctx->result, "error");
+    if (json_is_string(tmp)) {
+        if (strcmp(json_string_value(tmp), "authorization_pending") == 0) {
+            json_decref(dc_ctx->result);
+            return EAGAIN;
+        } else {
+            *error_description = get_json_string(dc_ctx, dc_ctx->result,
+                                                 "error_description");
+            DEBUG(SSSDBG_OP_FAILURE, "Token request failed with [%s][%s].\n",
+                                     json_string_value(tmp),
+                                     *error_description);
+            json_decref(dc_ctx->result);
+            return EIO;
+        }
+    }
+
+    /* Looks like we got the tokens */
+    dc_ctx->access_token = json_object_get(dc_ctx->result, "access_token");
+    dc_ctx->access_token_str = get_json_string(dc_ctx, dc_ctx->result,
+                                               "access_token");
+    dc_ctx->id_token = json_object_get(dc_ctx->result, "id_token");
+    dc_ctx->id_token_str = get_json_string(dc_ctx, dc_ctx->result, "id_token");
+
+    return EOK;
+}
+
+static const char *get_id_string(TALLOC_CTX *mem_ctx, json_t *id_object)
```

Comment:
It's unclear (inconsistent) if user of this function is expected to free memory.